src/ledger/test/LedgerTxnTests.cpp

```diff
@@ -1605,6 +1605,84 @@ TEST_CASE_VERSIONS("LedgerTxn load", "[ledgertxn]")
 #endif
 }
 
+TEST_CASE("LedgerTxn thread invariance", "[ledgertxn]")
+{
+    VirtualClock clock;
+    auto cfg = getTestConfig(0, Config::TESTDB_POSTGRESQL);
+    cfg.EXPERIMENTAL_PARALLEL_LEDGER_APPLY = true;
+    auto app = createTestApplication(clock, cfg);
+    LedgerTxn ltx1(app->getLedgerTxnRoot());
+    std::exception_ptr exception = nullptr;
+    auto runThreadTest = [&](std::function<void()> func,
+                             std::string const& msg) {
+        std::thread t([&]() {
+            try
+            {
+                func();
+            }
+            catch (...)
+            {
+                exception = std::current_exception();
+            }
+        });
+        t.join();
+        REQUIRE(exception);
+        REQUIRE_THROWS_WITH(std::rethrow_exception(exception), msg);
+    };
+
+    SECTION("LedgerTxnRoot does not allow different threads to create children")
+    {
+        SECTION("ltx1 is active")
+        {
+            runThreadTest([&]() { LedgerTxn ltx2(app->getLedgerTxnRoot()); },
+                          "ThreadInvariant called from wrong thread");
+        }
+        SECTION("ltx1 is inactive")
+        {
+            ltx1.rollback();
+            std::thread t([&]() {
+                // Ltx2 is ok, because ltx1 was reset
+                LedgerTxn ltx2(app->getLedgerTxnRoot());
+            });
+            t.join();
+        }
+    }
+    SECTION("re-using LedgerTxn across different threads")
+    {
+        std::string ledgerTxError = "LedgerTxn is accessed from wrong thread";
+        SECTION("create child")
+        {
+            runThreadTest([&]() { LedgerTxn ltx2(ltx1); }, ledgerTxError);
+        }
+        // `commit` and `rollback` terminate the program on an exception,
+        // because they are marked noexcept SECTION("commit child")
```

Comment:
This bit is somewhat worrying. Or rather: I wonder if the thread-invariant error ought to just be an abort in all cases, rather than an exception? the ltx is exception safe _because_ transactions run in an exception-catch block and use exceptions for "meaningfully recoverable" errors. But a thread-safety violation isn't really recoverable, right? It should probably always be fatal?

Reply:
Good point, abort is certainly more appropriate here.

Reply:
Unfortunately, in the process of landing all the p23 changes, this PR bit-rotted quite a bit. Some cleanups were already done in other PRs. I've extracted LedgerTxn change into a separate [PR](https://github.com/stellar/stellar-core/pull/4910), that we can merge independently of Herder changes.